lib/semantics/resolve-names.cc

```diff
@@ -4476,6 +4473,94 @@ const parser::Name *DeclarationVisitor::FindComponent(
   return nullptr;
 }
 
+// C764, C765
+void DeclarationVisitor::CheckInitialDataTarget(
+    const Symbol &pointer, const SomeExpr &expr) {
+  if (auto base{evaluate::GetBaseObject(expr)}) {
+    if (const Symbol * baseSym{base->symbol()}) {
+      const Symbol &ultimate{baseSym->GetUltimate()};
+      if (!IsAllocatable(ultimate) && ultimate.Corank() == 0 &&
+          ultimate.attrs().HasAll({Attr::TARGET, Attr::SAVE}) &&
+          evaluate::IsInitialDataTarget(expr)) {
+        // TODO: check type compatibility
+        // TODO: check non-deferred type parameter values
+        // TODO: check contiguity if pointer is CONTIGUOUS
+        if (pointer.Rank() != expr.Rank()) {
+          Say(pointer.name(),
+              "Pointer '%s' initialized with target of different rank"_err_en_US);
+        }
+        return;
+      }
+    }
+  }
+  Say(pointer.name(),
+      "Pointer '%s' initialized with invalid data target designator"_err_en_US);
```

Comment:
I think we should say why the initial data target is invalid, like with Rank above. This doesn't give the user much to go on.

Also, the message should be associated with the source location of the designator. In the example below the error is associated with the first line (because that's where the symbol for `p` is first created); it should be the third.
```
pointer :: p
real :: x
real :: p => x
end
```

Reply:
Done, with a test.

Reply:
Looks good, thanks.